test/common/common/mutex_tracer_test.cc

```diff
@@ -0,0 +1,108 @@
+#include <chrono>
+#include <thread>
+
+#include "common/common/lock_guard.h"
+#include "common/common/mutex_tracer.h"
+
+#include "test/test_common/test_time.h"
+
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class MutexTracerTest : public testing::Test {
+protected:
+  void SetUp() override {
+    tracer_ = MutexTracer::getOrCreateTracer();
+    tracer_->reset();
+  }
+
+  // Since MutexTracer::contentionHook is a private method, MutexTracerTest is a friend class.
+  void sendWaitCyclesToContentionHook(int64_t wait_cycles) {
+    tracer_->contentionHook(nullptr, nullptr, wait_cycles);
+  }
+
+  Thread::Thread launchThread() {
+    return Thread::Thread([this]() -> void { holdUntilContention(); });
+  }
+
+  Thread::MutexBasicLockable mu_;
+  MutexTracer* tracer_;
+  DangerousDeprecatedTestTime test_time_;
+
+private:
+  void holdUntilContention() {
+    int64_t curr_num_contentions = tracer_->numContentions();
+    while (tracer_->numContentions() == curr_num_contentions) {
+      test_time_.timeSystem().sleep(std::chrono::milliseconds(1));
```

Comment:
just comment that we hold the lock 90% of the time to ensure both contention and eventual acquisition, which is needed to bump numContentions().